### PR TITLE
Paris Law (Growth Crack Model)

### PR DESCRIPTION
**Pull request by hlembo**
A Simple Implementation of Paris Law.

### CODE REVIEW

File: examples/growth.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
+# Copyright © 2021 United States Government as represented by the Administrator of the
+# National Aeronautics and Space Administration.  All Rights Reserved.
+
+"""
+Example of a battery being simulated for a set period of time and then till threshold is met.
```
Comment (teubert) [inline examples/growth.py:5]: Update description here

File: examples/growth.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+"""
+
+from prog_models.models.paris_law import Growth 
+import matplotlib.pyplot as plt
+import csv
+import os
+# VVV Uncomment this to use Electro Chemistry Model VVV
+# from prog_models.models import BatteryElectroChem as Battery
```
Comment (teubert) [inline examples/growth.py:13]:
```suggestion
```

These lines are from the previous example

File: examples/growth.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            k_max = 24
+        elif (t < 750):
+            k_min = 8
+            k_max = 32
+        else:
+            k_min = 0
+            k_max = 28
+        return m.InputContainer({'k_min': k_min, 'k_max': k_max})
+
+
```
Comment (teubert) [inline examples/growth.py:34]:
```suggestion
    # Step 3: Estimate parameters
    # We do not know the model parameters for this system, 
    # so we need to estimate it using data collected from the system
    # First we have to import some data from the real system
    # This is what we use to estimate parameters
```

File: examples/growth.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            k_max = 28
+        return m.InputContainer({'k_min': k_min, 'k_max': k_max})
+
+    
+    times = []
+    inputs = []
+    outputs = []
+
+    #Finds file path
+
```
Comment (teubert) [inline examples/growth.py:40]:
```suggestion
```

File: examples/growth.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    print(' Error: ', m.calc_error(times, inputs, outputs, dt=10))
+
+    m.estimate_params([(times, inputs, outputs)], keys, dt=10)
+
+    print('\nOptimized configuration')
+    for key in keys:
+        print("-", key, m.parameters[key])
+    print(' Error: ', m.calc_error(times, inputs, outputs, dt=10))
+
+    # Simulate to threshold
```
Comment (teubert) [inline examples/growth.py:69]:
```suggestion
    # Step 4: Simulate to threshold
```

File: examples/growth.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        'save_freq': 10, # Frequency at which results are saved
+        'dt': 10, # Timestep
+        'print': True,
+        'progress': False,
+        'horizon': 1e5, # Horizon
+    }
+
+    (times, inputs, states, outputs, event_states) = m.simulate_to_threshold(future_loading, **options)
+
+    # Step 4: Plot Results
```
Comment (teubert) [inline examples/growth.py:82]:
```suggestion
    # Step 5: Plot Results
```

File: examples/growth.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    print(' Error: ', m.calc_error(times, inputs, outputs, dt=10))
+
+    # Simulate to threshold
+    print('\n\n------------------------------------------------')
+    print('Simulating to threshold\n\n')
+    options = {
+        'save_freq': 10, # Frequency at which results are saved
+        'dt': 10, # Timestep
+        'print': True,
+        'progress': False,
```
Comment (teubert) [inline examples/growth.py:76]:
```suggestion
```

It's false by default

File: examples/growth.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+    # Step 4: Plot Results
+    # crack length
+    # plot event state
+
+    inputs.plot(ylabel='Stress Intensity')
+    event_states.plot(ylabel= 'CGF')
+    outputs.plot(ylabel= {'c_li': "Crack Length"}, compact= False)
+    plt.show()
+if __name__ == '__main__':
```
Comment (teubert) [inline examples/growth.py:90]:
```suggestion

if __name__ == '__main__':
```

File: src/prog_models/models/paris_law.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+# INSTRUCTIONS:
+# 1. Copy this file- renaming to the name of your model
+# 2. Rename the class as desired
+# 3. Replace the events, inputs, states, outputs keys with those specific to the model
+# 4. Uncomment either dx or next_state function. dx for continuous models, and next_state for discrete
+# 5. Implement logic of model in each method
+
+# Note: To preserve vectorization use numpy math function (e.g., maximum, minimum, sign, sqrt, etc.) instead of non-vectorized functions (max, min, etc.)
+
```
Comment (teubert) [inline src/prog_models/models/paris_law.py:12]:
```suggestion
```

Remove instructions

File: src/prog_models/models/paris_law.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+# See examples.derived_params
+# 
+# Each function defines one or more derived parameters as a function of the other parameters.
+def example_callback(params):
+    # Return format: dict of key: new value pair for at least one derived parameter
+    return {
+        "Example Parameter 1": params["Example Parameter 2"]-3
+    }
+
+
```
Comment (teubert) [inline src/prog_models/models/paris_law.py:26]:
```suggestion
```

Remove callback (not used)

File: src/prog_models/models/paris_law.py [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+class Growth(PrognosticsModel):
```
Comment (teubert) [inline src/prog_models/models/paris_law.py:27]:
```suggestion
class CrackGrowth(PrognosticsModel): 
```
Lets name it crack growth to be a little more precise (we could be modeling growth of all sorts of things)

File: src/prog_models/models/paris_law.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+        "Example Parameter 1": params["Example Parameter 2"]-3
+    }
+
+
+class Growth(PrognosticsModel): 
+    """
+    Events: (1)
+        CGF :   Crack Growth Fracture 
+
+    # Correct
```
Comment (teubert) [inline src/prog_models/models/paris_law.py:32]:
```suggestion
```

File: src/prog_models/models/paris_law.py [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+def example_callback(params):
+    # Return format: dict of key: new value pair for at least one derived parameter
+    return {
+        "Example Parameter 1": params["Example Parameter 2"]-3
+    }
+
+
+class Growth(PrognosticsModel): 
+    """
+    Events: (1)
```
Comment (teubert) [inline src/prog_models/models/paris_law.py:29]: Could you add a description of the model before this line?

File: src/prog_models/models/paris_law.py [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+
+    # Correct
+    Inputs/Loading: (2)
+      |k_max :  Maximum crack growth 
+      |k_min :  Minimum crack growth 
+
+    # Work in Progress
+    # Meets two criteria expects to be changing over time for a system as your simulate it
+    # something you need to keep track of in order to calculate the event state or outputs
+    # and progress towards the event (Crack Growth Fracture)
```
Comment (teubert) [inline src/prog_models/models/paris_law.py:40]:
```suggestion
```

Remove your notes

File: src/prog_models/models/paris_law.py [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+       |a :     Length of crack
+       |c :     Constant
+       |m :     Constant
+       |n :     cycles per loading
+    """ 
+
+    # V Uncomment Below if the class is vectorized (i.e., if it can accept input to all functions as arrays) V
+    # is_vectorized = True
+
+    # REPLACE THE FOLLOWING LIST WITH EVENTS BEING PREDICTED
```
Comment (teubert) [inline src/prog_models/models/paris_law.py:57]:
```suggestion
```

Remove comments. 

Also remove some unneeded comments below (I don't want to mark each one)

File: src/prog_models/models/paris_law.py [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+        }
+
+       
+    # Note: Thresholds met equation below is not strictly necessary. By default threshold_met will check if event_state is ≤ 0 for each event
+    # crack length ends when we reach a configurable size
+    def threshold_met(self, x):
+        t_met = {
+           'CGF': x['c_li'] > self.parameters['config_length']
+        }
+        return t_met
```
Comment (teubert) [inline src/prog_models/models/paris_law.py:152]: Add a newline at the end of the file.

See: https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

**Review by teubert — CHANGES_REQUESTED** [review]
Logic looks good, just needs a little cleaning up.

File: examples/growth.py (hunk 1 above)
Comment (hlembo) [reply examples/growth.py]: Added

File: src/prog_models/models/paris_law.py (hunk 13 above)
Comment (hlembo) [reply src/prog_models/models/paris_law.py:13]: Added

**Review by teubert — APPROVED** [review]

**Pull request merged**